File: filosofos_java/src/hilos/Tenedor.java
package hilos;

import java.util.concurrent.locks.ReentrantLock;

/**
 * Clase que representa un tenedor en el problema de los filósofos comensales.
 * Utiliza un ReentrantLock para controlar el acceso exclusivo al tenedor.
 */
public class Tenedor {
    private final int id;
    private final ReentrantLock mutex;
    
    /**
     * Inicializa un tenedor.
     * 
     * @param id Identificador único del tenedor
     */
    public Tenedor(int id) {
        this.id = id;
        this.mutex = new ReentrantLock();
    }
    
    /**
     * Bloquea el mutex si el tenedor está libre.
     * Si el tenedor está siendo usado, el hilo se bloquea hasta que esté disponible.
     */
    public void tomar() {
        mutex.lock();
        System.out.println("  -> Tenedor " + id + " tomado");
    }
    
    /**
     * Desbloquea el mutex, liberando el tenedor para que otros lo puedan usar.
     */
    public void soltar() {
        mutex.unlock();
        System.out.println("  -> Tenedor " + id + " liberado");
    }
    
    /**
     * Obtiene el identificador del tenedor.
     * 
     * @return ID del tenedor
     */
    public int getId() {
        return id;
    }
}

File: filosofos_python/solucion_hilos/tenedor.py
import threading


class Tenedor:
    """
    Clase que representa un tenedor en el problema de los filósofos comensales.
    Utiliza un mutex para controlar el acceso exclusivo al tenedor.
    """
    
    def __init__(self, id: int):
        """
        Inicializa un tenedor.
        
        Args:
            id: Identificador único del tenedor
        """
        self.id = id
        self.mutex = threading.Lock()
    
    def tomar(self):
        """
        Bloquea el mutex si el tenedor está libre.
        Si el tenedor está siendo usado, el hilo se bloquea hasta que esté disponible.
        """
        self.mutex.acquire()
        print(f"  -> Tenedor {self.id} tomado")
    
    def soltar(self):
        """
        Desbloquea el mutex, liberando el tenedor para que otros lo puedan usar.
        """
        self.mutex.release()
        print(f"  -> Tenedor {self.id} liberado")
